File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int prev_prod = 1;
        int prod = 1;
        int maxm = nums[0];
        for(int n: nums) {
            prod = prod*n;
            if(n==0) {
                prev_prod = 1;
                prod = 1;
                maxm = Math.max(0,maxm);
                continue;
            }
            else if(prod<0) {
                // System.out.println(prev_prod + " " + prod);
                maxm = Math.max(prod/prev_prod,maxm);
                prev_prod = prev_prod==1?prod:Math.max(prod,prev_prod);
            }
            maxm = Math.max(prod,maxm);
            // System.out.println(maxm + " " + prod);
            
        }
        return maxm;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int maxm = INT_MIN;
        int prod = 1;
        int i,n = nums.size();
        int prev_prod = 1;
        for(i = 0; i <n ; i++){
            prod = prod * nums[i];
            
            // cout<<prev_prod<<" "<<prod<<endl;
            if(nums[i] == 0){
                prod = 1;
                prev_prod = 1;
                maxm = max(maxm,0);
            }
            else if(prod<0){
                maxm = max(maxm, prod/prev_prod);
                prev_prod = (prev_prod==1)?prod:max(prev_prod,prod);
            }
            else{
                maxm = max(maxm,prod);
            }
        }
        return maxm;
    }
};